stan/math/prim/fun/csr_u_to_z.hpp

```diff
@@ -19,12 +19,12 @@ namespace math {
  * @param[in] u U vector.
  * @param[in] i Index into resulting z vector.
  * @return z[i] where z is conversion from u.
- * @throw std::domain_error if u is zero length.
+ * @throw std::domain_error if u does not contain at least 2 elements.
  * @throw std::out_of_range if i is out of range.
  */
 inline int csr_u_to_z(const std::vector<int>& u, int i) {
-  check_positive("csr_u_to_z", "u.size()", u.size());
-  check_range("csr_u_to_z", "i", u.size(), i + 1, "index out of range");
+  check_greater("csr_u_to_z", "u.size()", u.size(), 1);
+  check_range("csr_u_to_z", "i", u.size() - 1, i + 1, "index out of range");
```

Comment:
I'm having trouble understanding why `u.size()` is wrong. I'm not very familiar with sparse matrix formats so I'll write out my logic. Tell me where I'm wrong.

If there are `M` rows, there are `M + 1` elements of `u.size()`.

Since `u` marks the beginnings of each of the rows (with a one-past-the-end index in the last element), this function appears to, for a given row `i`, compute the number of elements in that row.

Since `i` is a row number, `i < M`, and so `i + 1 < M + 1` and so as long as `i` is a row number no accesses are out of range.

The check as coded here is `i + 1 < M`. I think it should be `i + 1 < M + 1` or `i < M`.

Reply:
Srry it took me a sec to see this

> I'm having trouble understanding why u.size() is wrong. I'm not very familiar with sparse matrix formats so I'll write out my logic. Tell me where I'm wrong.

idt it's sparse matrices. The accessible range of a container is `[0, ..., N - 1]` since it's zero indexed. Then it's `i+1` because we access that `u[i + 1]` right?
```
 check_range("csr_u_to_z", "i", u.size() - 1, i + 1, "index out of range");
  return u[i + 1] - u[i];
```

Reply:
Well the `i + 1` didn't change. The `u.size()` -> `u.size() - 1` changed.

Reply:
Wouldn't x[x.size()] be a bug for any indexing? This sounds like a bug fix more than anything

Reply:
This check:

```
check_range("csr_u_to_z", "i", u.size() - 1, i + 1, "index out of range");
```

is checking if `i + 1 < u.size() - 1`.

I think `i` is always less than `u.size()` from the calling code. I don't know what `z` is supposed to be though which is why I'm having trouble debugging this code and I saw you had this pull open: https://github.com/stan-dev/math/pull/1844

Do you know what `z` is? This code is from the sparse matrix stuff.

Reply:
> This sounds like a bug fix more than anything

Yeah as part of this pull I think this is just a bug fix. But there isn't an issue in the pull description.

Reply:
Anyway the rest of the pull is great so I wanted to figure out what this is cause I assume it's probably here for a reason and shouldn't be tossed.

Reply:
idt we should not check this because the places we currently call it from do checks. That could change! `operator[]` for std vectors doesn't do a bounds check to verify `i + 1` is valid so I think it's reasonable to check that `i + 1` is within the indexing range of `[0, ..., N - 1]`

```cpp
inline int csr_u_to_z(const std::vector<int>& u, int i) {
  check_greater("csr_u_to_z", "u.size()", u.size(), 1);
  // check_range("csr_u_to_z", "i", u.size() - 1, i + 1, "index out of range");
  // below would grab out of bounds memory if (i + 1) > (u.size() - 1)
  return u[i + 1] - u[i];
}
```

> I don't know what z is supposed to be though which is why I'm having trouble debugging this code and I saw you had this pull open: #1844

I'm not following what in #1844 is related to this.

> Do you know what z is? This code is from the sparse matrix stuff.

I've never used this code but if I had to guess it looks like it's grabbing part of the row index for the next nonzero element. This is probably used in `rstanarm` so @bgoodri may be better to ask about this

Reply:
> I'm not following what in #1844 is related to this

Oh, the title of that pull is generalize u to z. This is csr_u_to_z. I didn't actually look through the source in that pull. These are unrelated and these are different us and zs I guess.

@bgoodri help

Reply:
lol icic. The code here looks reasonable to me tho'. What situation do you expect this to fail (or erroneously fail)?

Reply:
I think u is of length M + 1 so the right check is `i + 1 < M + 1` unless there is something about `u` and `z` I don't understand (likely).

From our sparse matrix * vector docs:
```
 *    - u: index of where each row starts in w,  length
 *      is equal to the number of rows plus one.  Last entry is
 *      one-past-the-end in w, following the Eigen spec.
```

I just don't know what `z` is or what it's trying to do lol.

Reply:
>  This is probably used in `rstanarm` so @bgoodri may be better to ask about this

We don't actually use any of the `csr_*` functions. We had to write our own to avoid doing all these checks on sparse matrices that never change.

Reply:
Hmm, I guess I'll just take another look at this and if I can't figure it out I'll just put the u_to_z stuff in another pull.

Reply:
I talked to Ben and I think it's fine to just change the range check to be `u.size()`. I forgot that the check is a less than and not a less than or equal to.

Reply:
I went digging with some `std::cout` s. Looks like `stan::error_index::value` is 1 by default, and so by default `check_range` is  treating things as if they are one-indexed. So I think this change is right.

Reply:
Indeed I approve